File: advent2017/extras/cxx/src/days/day01.cpp
#include <filesystem>
#include <fstream>
#include <chrono>
#include <cctype>
#include <iomanip>
#include <iostream>
#include <openssl/sha.h>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

const std::string kExpectedSha = "ffefe22d570c7077ac45df89cd8a40c99990e1903f6e68a501d75e53038c80ef";
const std::string kExpectedPart1 = "1158";
const std::string kExpectedPart2 = "1132";

std::string resolve_input(const std::string& provided) {
    if (!provided.empty()) return provided;
    const std::vector<std::string> cands{
        "advent2017/Day1/d1_input.txt",
        "Day1/d1_input.txt",
        "../Day1/d1_input.txt",
        "../../Day1/d1_input.txt",
    };
    for (const auto& c : cands) {
        if (std::filesystem::exists(c)) return c;
    }
    throw std::runtime_error("input not found");
}

std::string read_all(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) throw std::runtime_error("failed to read input");
    return std::string((std::istreambuf_iterator<char>(in)), {});
}

std::string sha256_hex(const std::string& data) {
    unsigned char out[SHA256_DIGEST_LENGTH];
    SHA256(reinterpret_cast<const unsigned char*>(data.data()), data.size(), out);
    std::ostringstream oss;
    oss << std::hex << std::setfill('0');
    for (auto b : out) {
        oss << std::setw(2) << static_cast<int>(b);
    }
    return oss.str();
}

std::string trim_ws(std::string s) {
    while (!s.empty() && std::isspace(static_cast<unsigned char>(s.back()))) s.pop_back();
    std::size_t i = 0;
    while (i < s.size() && std::isspace(static_cast<unsigned char>(s[i]))) ++i;
    return s.substr(i);
}

int solve_with_step(const std::string& raw, int step) {
    const std::string s = trim_ws(raw);
    if (s.empty()) return 0;
    int sum = 0;
    const int n = static_cast<int>(s.size());
    for (int i = 0; i < n; ++i) {
        const int j = (i + step) % n;
        if (s[i] == s[j]) {
            sum += (s[i] - '0');
        }
    }
    return sum;
}

}  // namespace

int main(int argc, char** argv) {
    int part = 0;
    std::string input;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--part") part = std::stoi(argv[++i]);
        else if (a == "--input") input = argv[++i];
        else throw std::runtime_error("unknown arg");
    }
    if (part != 1 && part != 2) throw std::runtime_error("--part 1|2 required");

    input = resolve_input(input);
    const std::string raw = read_all(input);
    if (sha256_hex(raw) != kExpectedSha) throw std::runtime_error("checksum mismatch");

    auto t0 = std::chrono::steady_clock::now();
    std::string ans;
    std::string expected;
    if (part == 1) {
        ans = std::to_string(solve_with_step(raw, 1));
        expected = kExpectedPart1;
    } else {
        const int n = static_cast<int>(trim_ws(raw).size());
        ans = std::to_string(solve_with_step(raw, n / 2));
        expected = kExpectedPart2;
    }
    if (ans != expected) throw std::runtime_error("answer mismatch");

    std::cout << ans << "\n";
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::cerr << "[cxx-fancy] day=1 part=" << part << " runtime_ms=" << ms << "\n";
    return 0;
}

File: advent2017/extras/rust/src/bin/day01.rs
use sha2::{Digest, Sha256};
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "ffefe22d570c7077ac45df89cd8a40c99990e1903f6e68a501d75e53038c80ef";
const EXPECTED_PART1: &str = "1158";
const EXPECTED_PART2: &str = "1132";

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() {
        return provided.to_string();
    }
    let cands = [
        "advent2017/Day1/d1_input.txt",
        "Day1/d1_input.txt",
        "../Day1/d1_input.txt",
        "../../Day1/d1_input.txt",
    ];
    for c in cands {
        if Path::new(c).exists() {
            return c.to_string();
        }
    }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new();
    h.update(&b);
    hex::encode(h.finalize())
}

fn solve_with_step(raw: &str, step: usize) -> i64 {
    let s = raw.trim().as_bytes();
    if s.is_empty() {
        return 0;
    }
    let mut total: i64 = 0;
    for i in 0..s.len() {
        let j = (i + step) % s.len();
        if s[i] == s[j] {
            total += (s[i] - b'0') as i64;
        }
    }
    total
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => { part = args[i + 1].parse().unwrap(); i += 2; }
            "--input" => { input = args[i + 1].clone(); i += 2; }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 { panic!("--part 1|2 required") }

    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA {
        panic!("checksum mismatch");
    }
    let raw = fs::read_to_string(&in_path).expect("read_to_string");

    let t0 = Instant::now();
    let (ans, expected) = if part == 1 {
        (solve_with_step(&raw, 1).to_string(), EXPECTED_PART1)
    } else {
        let n = raw.trim().len();
        (solve_with_step(&raw, n / 2).to_string(), EXPECTED_PART2)
    };

    if ans != expected {
        panic!("answer mismatch");
    }
    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=1 part={} runtime_ms={:.3}", part, ms);
}
